source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -553,6 +553,124 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
   return status;
 }
 
+FilterDataStatus Filter::handleDataBufferedMode(ProcessorState& state, Buffer::Instance& data,
+                                                bool end_stream) {
+  if (end_stream) {
+    switch (openStream()) {
+    case StreamOpenState::Error:
+      return FilterDataStatus::StopIterationNoBuffer;
+    case StreamOpenState::IgnoreError:
+      return FilterDataStatus::Continue;
+    case StreamOpenState::Ok:
+      // Fall through
+      break;
+    }
+
+    // The body has been buffered and we need to send the buffer
+    ENVOY_LOG(debug, "Sending request body message");
+    state.addBufferedData(data);
+    auto req = setupBodyChunk(state, *state.bufferedData(), end_stream);
+    sendBodyChunk(state, ProcessorState::CallbackState::BufferedBodyCallback, req);
+    // Since we just just moved the data into the buffer, return NoBuffer
+    // so that we do not buffer this chunk twice.
+    state.setPaused(true);
+    return FilterDataStatus::StopIterationNoBuffer;
+  }
+  ENVOY_LOG(trace, "onData: Buffering");
+  state.setPaused(true);
+  return FilterDataStatus::StopIterationAndBuffer;
+}
+
+FilterDataStatus Filter::handleDataStreamedMode(ProcessorState& state, Buffer::Instance& data,
+                                                bool end_stream) {
+  // STREAMED body mode works as follows:
+  //
+  // 1) As data callbacks come in to the filter, it "moves" the data into a new buffer, which it
+  // dispatches via gRPC message to the external processor, and then keeps in a queue. It
+  // may request a watermark if the queue is higher than the buffer limit to prevent running
+  // out of memory.
+  // 2) As a result, filters farther down the chain see empty buffers in some data callbacks.
+  // 3) When a response comes back from the external processor, it injects the processor's result
+  // into the filter chain using "inject**codedData". (The processor may respond indicating that
+  // there is no change, which means that the original buffer stored in the queue is what gets
+  // injected.)
+  //
+  // This way, we pipeline data from the proxy to the external processor, and give the processor
+  // the ability to modify each chunk, in order. Doing this any other way would have required
+  // substantial changes to the filter manager. See
+  // https://github.com/envoyproxy/envoy/issues/16760 for a discussion.
+  switch (openStream()) {
+  case StreamOpenState::Error:
+    return FilterDataStatus::StopIterationNoBuffer;
+  case StreamOpenState::IgnoreError:
+    return FilterDataStatus::Continue;
+  case StreamOpenState::Ok:
+    // Fall through
```

Comment:
ABSL_FALLTHROUGH

Reply:
I don't think we want to add ABSL_FALLTHROUGH_INTENDED here.
Do you mean add it at line 721?

Reply:
oh, sorry. The "fall through" comment confused me. We don't need one here.

I am wondering what "fall through" means before a 'break'.

Reply:
Yeah, it's obvious. But, let's just leave it there.
Thanks for approval.  Could you please merge it so I can perform an upstream merge for PR: https://github.com/envoyproxy/envoy/pull/34942?